_posts/2020-07-08-event-driven-functions-with-openfaas-and-nats.md

```diff
@@ -0,0 +1,146 @@
+dark_background: true
+---
+
+In this blog post, I will show how you can invoke OpenFaaS function in response to messages sent on NATS topics in publish-subscribe model.
+OpenFaaS functions are accessible over HTTP endpoints via gateway service but OpenFaaS provides several other way to invoke OpenFaaS functions with help of [connector-sdk](https://github.com/openfaas-incubator/connector-sdk). 
+
+* [kafka-connector](https://github.com/openfaas-incubator/kafka-connector) connects OpenFaaS functions to Kafka topics.
+* [nats-connector](https://github.com/openfaas-incubator/nats-connector) an OpenFaaS event-connector to trigger functions from NATS. 
+* [mqtt-connector](https://github.com/openfaas-incubator/mqtt-connector) MQTT connector for OpenFaaS.
+* [cron-connector](https://github.com/openfaas-incubator/cron-connector) triggers OpenFaaS functions based on cron events. 
+* [VMware vCenter connector](https://github.com/openfaas-incubator/openfaas-vcenter-connector) an OpenFaaS event-connector built to consume events from vCenter and to trigger functions.
+
+There are several other connectors which allows you to trigger OpenFaaS functions based on events, are written and managed as a third party project. Please refer to this [link](https://docs.openfaas.com/reference/triggers/) for full list.
+
+## NATS
+
+[NATS](https://nats.io) is a simple, secure and high performance open source messaging system for cloud native applications, IoT messaging, and microservices architectures. In this blog post, I will be using NATS publish-subscribe concept where publishers publishes a message on a topic/subject and subscribers consumes that message by subscribing to that topic/subject.
+
+![NATS Publish Subscribe](/images/2020-openfaas-nats/nats-publish-subscribe.png "https://docs.nats.io/nats-concepts/pubsub")
+
+## Pre-requisite Installation
+
+* [k3d](https://github.com/rancher/k3d) to create a [k3s](https://github.com/rancher/k3s) cluster. k3s is lightweight distribution of kubernetes from Rancher Labs.
+* `kubectl` to manage kubernetes cluster.
+* `helm` to install `nats-connector` using helm chart
+* [arkade](https://github.com/alexellis/arkade) to install OpenFaaS. `arkade` is simple CLI tool to install helm charts and apps to your cluster in one command.
+* [faas-cli](https://github.com/openfaas/faas-cli) CLI tool to manage OpenFaaS functions.
+
+> Please make sure you have these tools installed before you proceed next.
+
+## Install Components
+
+#### Create a kubernetes cluster
+
+`k3d` is a tool to run `k3s` in docker. Create a single node `k3s` cluster.
+
+```
+k3d create
+```
+
+Follow instructions returned in the output of this command to configure KUBECONFIG to access the cluster using kubectl.
+
+#### Install OpenFaaS 
+Install OpenFaaS using `arkade`
+
+```
+arkade install openfaas --set basic_auth=true --set functionNamespace=openfaas-fn
+```
+
+Follow instructions printed to obtain the gateway password and login with `faas-cli`
```

Comment:
there can be a lot of output in the  terminal, given the installation with arkade, i think it is safe to include the login command here as well, but @alexellis might have a different opinion

Reply:
Added the command below

Reply:
I am conflicted about this - in one sense repeating it makes it obvious, in another if it every changes, then it will be a nightmare to update it everywhere on the web.